app/code/Magento/Sitemap/Plugin/StoreEmulation.php

```diff
@@ -0,0 +1,60 @@
+<?php
+/**
+* Copyright (c) 2020 TechDivision GmbH <info@techdivision.com> - TechDivision GmbH
+* All rights reserved
+*
+* This product includes proprietary software developed at TechDivision GmbH, Germany
+* For more information see http://www.techdivision.com/
+*
+* To obtain a valid license for using this software please contact us at
+* license@techdivision.com
+*/
+
+namespace Magento\Sitemap\Plugin;
+
+use Magento\Sitemap\Model\Sitemap;
+use Magento\Store\Model\App\Emulation;
+
+/**
+* @copyright  Copyright (c) 2020 TechDivision GmbH <info@techdivision.com> - TechDivision GmbH
+*
+* @link       https://www.techdivision.com/
+* @author     CSI-Team <csi-kolbermoor@techdivision.com >
+*/
+class StoreEmulation
+{
+   /** @var Emulation $appEmulation */
+   private $appEmulation;
+
+   /**
+    * @param Emulation $appEmulation
+    */
+   public function __construct(Emulation $appEmulation)
+   {
+       $this->appEmulation = $appEmulation;
+   }
+
+   /**
+    * @param Sitemap $subject
+    * @see \Magento\Sitemap\Model\Sitemap::generateXml
+    */
+   public function beforeGenerateXml(Sitemap $subject)
```

Comment:
Better to use around plugin here
and write code in the way like 
```
 $this->appEmulation->startEnvironmentEmulation(
                    $sitemap->getStoreId(),
                    Area::AREA_FRONTEND,
                    true
                );
             try {
                retrun $proceed();
             } finaly {
                $this->appEmulation->stopEnvironmentEmulation();
             }
```

Reply:
And actually why do not put this directly in generateXml method

Reply:
> And actually why do not put this directly in generateXml method

@kandy why do you think Sitemap itself should be responsible for modifying the Application state? I'd say it should be done by the observer.